### ISSUE TITLE
epoch数量

### ISSUE THREAD

**Issue by qingqing144545** [musicman217/Text-Proxy#3]
作者你好，我复现了你的代码，但我发现，为什么论文里面提到MSRVTT只训练5个epoch，但实验却设置了训练100个epoch？
“We optimize the model for 5 epochs for MSRVTT on 2 RTX 4090 GPUs”

"num_train_epochs": 100,
  "min_valid_steps": 1,
  "num_valid": 1,
  "only_valid_steps": 100,
  "save_steps_ratio": 0.9,
  "output_dir": "vidclip_data/output/msrvtt_retrieval/msrvtt_retrieval_vip_base_32",
  "if_tb_log": 0,
  "if_model_saver": 1,
  "if_log2file": 1,
  "dummy_data": 0

**Comment by musicman217** [comment]
> 作者你好，我复现了你的代码，但我发现，为什么论文里面提到MSRVTT只训练5个epoch，但实验却设置了训练100个epoch？ “We optimize the model for 5 epochs for MSRVTT on 2 RTX 4090 GPUs”
> 
> "num_train_epochs": 100, "min_valid_steps": 1, "num_valid": 1, "only_valid_steps": 100, "save_steps_ratio": 0.9, "output_dir": "vidclip_data/output/msrvtt_retrieval/msrvtt_retrieval_vip_base_32", "if_tb_log": 0, "if_model_saver": 1, "if_log2file": 1, "dummy_data": 0

您好，这个cfg设定也是按照baseline CLIPViP来的，我并没做更改，这个num_train_epochs参数并非所谓的epoch，你可以看下main脚本中训练总step数量的定义，当整体batchsize为128时，总共的step数量是7030，这和其他基于CLIP4Clip框架下的设定是一样的，总共是5轮epoch

**Comment by qingqing144545** [comment]
作者你好，非常感谢您的回复，但是我还有点不懂向您请教，因为msrvtt训练集每个v有20个t，基于clip4clip的baseline的每个epoch是9000*20/32=5625，5625*5=28125；然后我想的是clipvip是9000/128*100=7032，然后因为batchsize4倍的关系，最后总步数也确实是4倍，所以这个100确实并非所谓的epoch，但是为什么这样设计向您请教，还有clipvip里这个100是分成5个20，每20个num_train_epochs才算真正的epoch吗？还有就是作者您可以提供只推理而不训练的命令脚本吗？

**Comment by musicman217** [comment]
> 作者你好，非常感谢您的回复，但是我还有点不懂向您请教，因为msrvtt训练集每个v有20个t，基于clip4clip的baseline的每个epoch是9000_20/32=5625，5625_5=28125；然后我想的是clipvip是9000/128*100=7032，然后因为batchsize4倍的关系，最后总步数也确实是4倍，所以这个100确实并非所谓的epoch，但是为什么这样设计向您请教，还有clipvip里这个100是分成5个20，每20个num_train_epochs才算真正的epoch吗？还有就是作者您可以提供只推理而不训练的命令脚本吗？

你好，这份代码和CLIP4Clip框架不同的是，clip4clip是将全部180000（9000*20）个pair作为训练集，而clipvip代码中，实际对于给定某个video id，是从其对应的20个captions中随机抽取一个得到一个样本对，这个可以从dataloader的 __getitem__函数中得知，所以train dataloader的大小就是9000（个视频），只不过某个video id在同一个epoch下会在其20个captions中随机放回抽取20次作为当前epoch下该video的20个pair，而我们要进行5轮epoch，所以num_train_epochs = 20*5 = 100，最终我们模型更新的step数量还是7032（基于整体batch size=128时），所以这个num_train_epochs = 100 的含义可以理解为对于某个video来说模型会使用它所在的batch来进行100次更新

由于代码清洗的缘故我并没有保存单纯推理部分的脚本，你可以训练过程中保存最高recall值的模型权重，然后将run_video_retrieval.py中的training部分抹去，直接加载权重然后进行推理，需要注意的是，由于随机种子的设定，这可能会改变推理阶段的recall结果

**Comment by qingqing144545** [comment]
作者您好，感谢您之前的回复，对我理解数据加载很有帮助，我用两张A100跑了msrvtt数据集，在4400step，logits weight取0.4的时候达到了最好的效果如下

03/16/2025 17:55:11 - INFO - __main__ -   ==================  inference proxy logits weight : 0.4 =====================
03/16/2025 17:55:11 - INFO - __main__ -   Evaluate under setting: simple.
03/16/2025 17:55:11 - INFO - __main__ -   validation finished in 9 seconds, validated on 1000 videos 
msrvtt-1ka t2v recall@1: 52.0000 msrvtt-1ka t2v recall@5: 76.3000 msrvtt-1ka t2v recall@10: 84.4000 msrvtt-1ka t2v recall_med: 1.0 msrvtt-1ka t2v recall_mean: 12.7 msrvtt-1ka v2t recall@1: 50.3960 msrvtt-1ka v2t recall@5: 76.1386 msrvtt-1ka v2t recall@10: 84.8515 msrvtt-1ka v2t recall_med: 1.0 msrvtt-1ka v2t recall_mean: 9.6

比起你论文的数据差了一个点，然后我尝试只推理跑出的model_best.pt,但是效果却差强人意，recall1只有28.3（我去除了dsl）

/16/2025 19:43:38 - INFO - __main__ -   [msrvtt-9k] is_train True dataset size 9000,
03/16/2025 19:43:38 - INFO - __main__ -   [msrvtt-1ka] is_train False dataset size 1000,
03/16/2025 19:43:38 - INFO - __main__ -   [msrvtt-1ka] is_train False dataset size 1000,
03/16/2025 19:43:38 - INFO - __main__ -   🔹 Running inference only...
03/16/2025 19:43:38 - INFO - __main__ -   Using e2e_weights_path: /home/imuse/zmq_workspace/Text-Proxy/outputs32/ckpt/model_best.pt
03/16/2025 19:43:38 - INFO - __main__ -   Output directory: /home/imuse/zmq_workspace/Text-Proxy/outputs32
03/16/2025 19:43:38 - INFO - __main__ -   Loop val_loader msrvtt-1ka.
03/16/2025 19:43:53 - INFO - __main__ -   ==================  inference proxy logits weight : 0.0 =====================
03/16/2025 19:43:53 - INFO - __main__ -   Evaluate under setting: simple.
03/16/2025 19:43:53 - INFO - __main__ -   validation finished in 14 seconds, validated on 1000 videos
msrvtt-1ka t2v recall@1: 28.3000 msrvtt-1ka t2v recall@5: 50.0000 msrvtt-1ka t2v recall@10: 60.0000 msrvtt-1ka t2v recall_med: 5.5 msrvtt-1ka t2v recall_mean: 45.7 msrvtt-
1ka v2t recall@1: 24.9505 msrvtt-1ka v2t recall@5: 48.2178 msrvtt-1ka v2t recall@10: 59.4059 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.4
03/16/2025 19:43:53 - INFO - __main__ -   Best t2v recall: 0.283   
03/16/2025 19:43:53 - INFO - __main__ -   ==================  inference proxy logits weight : 0.1 =====================
03/16/2025 19:43:53 - INFO - __main__ -   Evaluate under setting: simple.
03/16/2025 19:43:53 - INFO - __main__ -   validation finished in 14 seconds, validated on 1000 videos
msrvtt-1ka t2v recall@1: 27.9000 msrvtt-1ka t2v recall@5: 50.0000 msrvtt-1ka t2v recall@10: 59.8000 msrvtt-1ka t2v recall_med: 5.5 msrvtt-1ka t2v recall_mean: 46.0 msrvtt-1ka v2t recall@1: 25.0495 msrvtt-1ka v2t recall@5: 48.1188 msrvtt-1ka v2t recall@10: 59.3069 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.5
03/16/2025 19:43:53 - INFO - __main__ -   Best t2v recall: 0.283
03/16/2025 19:43:53 - INFO - __main__ -   ==================  inference proxy logits weight : 0.2 =====================
03/16/2025 19:43:53 - INFO - __main__ -   Evaluate under setting: simple.
03/16/2025 19:43:53 - INFO - __main__ -   validation finished in 14 seconds, validated on 1000 videos
msrvtt-1ka t2v recall@1: 27.1000 msrvtt-1ka t2v recall@5: 48.7000 msrvtt-1ka t2v recall@10: 59.7000 msrvtt-1ka t2v recall_med: 6.0 msrvtt-1ka t2v recall_mean: 47.6 msrvtt-1ka v2t recall@1: 25.2475 msrvtt-1ka v2t recall@5: 47.4257 msrvtt-1ka v2t recall@10: 58.8119 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.8


请问我的修改是否有问题：我在run_video_retrieval.py285行后添加了如下判断

# **🔹 新增推理模式判断**
    if cfg.do_inference:
        LOGGER.info("🔹 Running inference only...")  
        LOGGER.info(f"Using e2e_weights_path: {cfg.e2e_weights_path}")
        inference_loaders = {k: PrefetchLoader(v, None) for k, v in inference_loaders.items()}
        validate(model, inference_loaders, cfg, device)  # 直接进行推理
        LOGGER.info(" Inference completed!")
        return  # **退出，防止进入训练流程**

这是我的test脚本：

export CUDA_VISIBLE_DEVICES=2,3
PROJ_PATH= [YOUR PROJECT DIRECTORY]
export PYTHONPATH=$PYTHONPATH:${PROJ_PATH}
python -m torch.distributed.launch --nproc_per_node=2 --master_port 29664 \
  src/tasks/run_video_retrieval.py \
  --config src/configs/msrvtt_retrieval/msrvtt_retrieval_vip_base_32.json \
  --blob_mount_dir ${PROJ_PATH} \
  --do_inference 1 \
  --e2e_weights_path /… /model_best.pt

**Comment by musicman217** [comment]
> 作者您好，感谢您之前的回复，对我理解数据加载很有帮助，我用两张A100跑了msrvtt数据集，在4400step，logits weight取0.4的时候达到了最好的效果如下
> 
> 03/16/2025 17:55:11 - INFO - **main** - ================== inference proxy logits weight : 0.4 ===================== 03/16/2025 17:55:11 - INFO - **main** - Evaluate under setting: simple. 03/16/2025 17:55:11 - INFO - **main** - validation finished in 9 seconds, validated on 1000 videos msrvtt-1ka t2v recall@1: 52.0000 msrvtt-1ka t2v recall@5: 76.3000 msrvtt-1ka t2v recall@10: 84.4000 msrvtt-1ka t2v recall_med: 1.0 msrvtt-1ka t2v recall_mean: 12.7 msrvtt-1ka v2t recall@1: 50.3960 msrvtt-1ka v2t recall@5: 76.1386 msrvtt-1ka v2t recall@10: 84.8515 msrvtt-1ka v2t recall_med: 1.0 msrvtt-1ka v2t recall_mean: 9.6
> 
> 比起你论文的数据差了一个点，然后我尝试只推理跑出的model_best.pt,但是效果却差强人意，recall1只有28.3（我去除了dsl）
> 
> /16/2025 19:43:38 - INFO - **main** - [msrvtt-9k] is_train True dataset size 9000, 03/16/2025 19:43:38 - INFO - **main** - [msrvtt-1ka] is_train False dataset size 1000, 03/16/2025 19:43:38 - INFO - **main** - [msrvtt-1ka] is_train False dataset size 1000, 03/16/2025 19:43:38 - INFO - **main** - 🔹 Running inference only... 03/16/2025 19:43:38 - INFO - **main** - Using e2e_weights_path: /home/imuse/zmq_workspace/Text-Proxy/outputs32/ckpt/model_best.pt 03/16/2025 19:43:38 - INFO - **main** - Output directory: /home/imuse/zmq_workspace/Text-Proxy/outputs32 03/16/2025 19:43:38 - INFO - **main** - Loop val_loader msrvtt-1ka. 03/16/2025 19:43:53 - INFO - **main** - ================== inference proxy logits weight : 0.0 ===================== 03/16/2025 19:43:53 - INFO - **main** - Evaluate under setting: simple. 03/16/2025 19:43:53 - INFO - **main** - validation finished in 14 seconds, validated on 1000 videos msrvtt-1ka t2v recall@1: 28.3000 msrvtt-1ka t2v recall@5: 50.0000 msrvtt-1ka t2v recall@10: 60.0000 msrvtt-1ka t2v recall_med: 5.5 msrvtt-1ka t2v recall_mean: 45.7 msrvtt- 1ka v2t recall@1: 24.9505 msrvtt-1ka v2t recall@5: 48.2178 msrvtt-1ka v2t recall@10: 59.4059 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.4 03/16/2025 19:43:53 - INFO - **main** - Best t2v recall: 0.283 03/16/2025 19:43:53 - INFO - **main** - ================== inference proxy logits weight : 0.1 ===================== 03/16/2025 19:43:53 - INFO - **main** - Evaluate under setting: simple. 03/16/2025 19:43:53 - INFO - **main** - validation finished in 14 seconds, validated on 1000 videos msrvtt-1ka t2v recall@1: 27.9000 msrvtt-1ka t2v recall@5: 50.0000 msrvtt-1ka t2v recall@10: 59.8000 msrvtt-1ka t2v recall_med: 5.5 msrvtt-1ka t2v recall_mean: 46.0 msrvtt-1ka v2t recall@1: 25.0495 msrvtt-1ka v2t recall@5: 48.1188 msrvtt-1ka v2t recall@10: 59.3069 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.5 03/16/2025 19:43:53 - INFO - **main** - Best t2v recall: 0.283 03/16/2025 19:43:53 - INFO - **main** - ================== inference proxy logits weight : 0.2 ===================== 03/16/2025 19:43:53 - INFO - **main** - Evaluate under setting: simple. 03/16/2025 19:43:53 - INFO - **main** - validation finished in 14 seconds, validated on 1000 videos msrvtt-1ka t2v recall@1: 27.1000 msrvtt-1ka t2v recall@5: 48.7000 msrvtt-1ka t2v recall@10: 59.7000 msrvtt-1ka t2v recall_med: 6.0 msrvtt-1ka t2v recall_mean: 47.6 msrvtt-1ka v2t recall@1: 25.2475 msrvtt-1ka v2t recall@5: 47.4257 msrvtt-1ka v2t recall@10: 58.8119 msrvtt-1ka v2t recall_med: 6.0 msrvtt-1ka v2t recall_mean: 41.8
> 
> 请问我的修改是否有问题：我在run_video_retrieval.py285行后添加了如下判断
> 
> # **🔹 新增推理模式判断**
> ```
> if cfg.do_inference:
>     LOGGER.info("🔹 Running inference only...")  
>     LOGGER.info(f"Using e2e_weights_path: {cfg.e2e_weights_path}")
>     inference_loaders = {k: PrefetchLoader(v, None) for k, v in inference_loaders.items()}
>     validate(model, inference_loaders, cfg, device)  # 直接进行推理
>     LOGGER.info(" Inference completed!")
>     return  # **退出，防止进入训练流程**
> ```
> 
> 这是我的test脚本：
> 
> export CUDA_VISIBLE_DEVICES=2,3 PROJ_PATH= [YOUR PROJECT DIRECTORY] export PYTHONPATH=$PYTHONPATH:${PROJ_PATH} python -m torch.distributed.launch --nproc_per_node=2 --master_port 29664 src/tasks/run_video_retrieval.py --config src/configs/msrvtt_retrieval/msrvtt_retrieval_vip_base_32.json --blob_mount_dir ${PROJ_PATH} --do_inference 1 --e2e_weights_path /… /model_best.pt

你好，我已经上传了eval的脚本，需要事先保存最高recall的权重至Text-Proxy/ 下为model_best.pth，只需要调用Text-Proxy/msrvrr_eval.sh即可，但需要注意的是，尽管在training stage的4400steps时进行validate能够打到recall@1=52.3的结果，直接使用该权重进行inference会有显著的性能倒退，我的结果为47上下，这可能是训练脚本和推理脚本的随机性导致，正如上面所讨论的，每个video是从20个captions中随机采样一个caption来组成样本对
